web3/datastructures.py

```diff
@@ -123,6 +123,26 @@ def __eq__(self, other: Any) -> bool:
             return False
 
 
+def tupleize_lists_nested(d: Mapping[TKey, TValue]) -> AttributeDict[TKey, TValue]:
+    """
+    Lists inside dicts will throw an error if attempted to be hashed.
+    This method converts them to tuples, rendering them hashable.
+    """
+
+    def _to_tuple(lst: List[Any]) -> Any:
+        return tuple(_to_tuple(i) if isinstance(i, list) else i for i in lst)
```

Comment:
I wonder if we should be less selective in general. I wouldn't expect too many `set` types to come in but those are also not hashable, as an example. Rather than make this utility method about `list` types exclusively, what do you think about targeting `Iterable` types in general if they are not tuples?

Something like

```py
def _is_non_tuple_iterable(value: Iterable[Any]) -> bool:
    return isinstance(value, Iterable) and not isinstance(value, tuple)


def dict_to_attrdict_hashable(d: Mapping[TKey, TValue]) -> AttributeDict[TKey, TValue]:
    """
    This method converts a ``dict`` to an ``AttributeDict`` while converting any
    un-hashable ``Iterable`` dict values to hashable ``tuple`` types.
    """

    ret = dict()
    for k, v in d.items():
        if isinstance(v, Mapping):
            ret[k] = dict_to_attrdict_hashable(v)
        elif _is_non_tuple_iterable(v):
            ret[k] = _recursive_iterable_to_tuple(v)
        else:
            ret[k] = v
    return AttributeDict(ret)
```

We can add some `set` examples in the tests too and the above should pass.

Thoughts?

Reply:
To me the changes look good and would help generalize the code. I also like the idea of checking for the generic `Mapping` type in particular. Thanks a lot @fselmo, learned some things.

My only concern/question: I had chosen the `list`-type approach because some of the logs I was retrieving from deployed Solidity contracts contained lists. Not an expert with Solidity/Vyper (yet), but would it ever be possible for a contract to return any other data type inside logs?
While I agree the structure is better, I can't say if the extensibility materializes in use cases and possibly could create surface for harder to spot issues (maybe increased memory consumption/hidden code execution due to 'auto-unrolling' of `Iterable` types? - just a thought).

Happy to add some tests with `set` or general `Iterable` types if we go with this approach. Thanks a lot for the input @pacrob  and @fselmo.

Reply:
I don't know what other datastructures could show up either, but I don't think it'll hurt to check. I like opening it up, but what about checking if it is iterable and not hashable, like:

```py
def _is_non_hashable_iterable(value: Any) -> bool:
    return isinstance(value, Iterable) and not isinstance(value, Hashable)

def _recursive_iterable_to_tuple(itr: Iterable[Any]) -> Any:
    return tuple(
        _recursive_iterable_to_tuple(i) if _is_non_hashable_iterable(i) else i
        for i in itr
    )

def mapping_to_attrdict_hashable(d: Mapping[TKey, TValue]) -> AttributeDict[TKey, TValue]:
    """
    This method converts a ``Mapping`` to an ``AttributeDict`` while converting any
    un-hashable ``Iterable`` dict values to hashable ``tuple`` types.
    """

    ret = dict()
    for k, v in d.items():
        if isinstance(v, Mapping):
            ret[k] = mapping_to_attrdict_hashable(v)
        elif _is_non_hashable_iterable(v):
            ret[k] = _recursive_iterable_to_tuple(v)
        else:
            ret[k] = v
    return AttributeDict(ret)
```

I've got a couple tests for `set` and `bytearray` passing locally. What other non-hashable iterables are there?

Reply:
Changes pushed. I changed the typing a bit, but I think it's an improvement.

Reply:
Nevermind my confidence - calling `tuple(set))` does not guarantee order, and `tuple(sorted(Iterable))` throws an error if the iterable contains non-comparable items (`int` and `tuple` in the test case).

A custom sorting function seems like overkill, but excluding `set`s from being in AttributeDicts also sounds complex.

Any preference or other ideas?

Reply:
Hmm not sure I understand the issue but I'm OK with just sticking with the original `list` -> `tuple` idea for now if it's too complicated. Perhaps that was the best way to go all along 🙃 . I think that way if any issue does come up we can tailor it to the use case a bit more.